feat: add all required endpoints

rohitxsh opened the pull request.

rohitxsh committed:
feat: add all required endpoints

rohitxsh committed:
chore: update readme

andrewyatz reviewed (approved):
As a first pass this looks good to me

marcoooo commented on main.py:
```diff
@@ -1,15 +1,77 @@
+AWS_SCHEMA_DATABASE_NAME = "ensembl-parquet-meta-schema"
+AWS_S3_OUTPUT_DIR = "s3://ensembl-athena-results/"
+
+@app.get("/api/v1")
```
You'd better think of your app as "behind a proxy" e.g. the relative uri is never mentioned in your code, so the day you switch to /api/v2 your routes definition don't change. More info here: https://fastapi.tiangolo.com/advanced/behind-a-proxy/
Therefore your route must be defined as "relative", this one should ten become
```suggestion
@app.get("/")
```

marcoooo commented on main.py:
```diff
@@ -1,15 +1,77 @@
+    return "Ensembl's data lakehouse backend"
+
+
+@app.get("/api/v1/data")
```
Following first comment this one should become
```suggestion
@app.get("/data")
```

marcoooo commented on main.py:
```diff
@@ -1,15 +1,77 @@
+    return query_response
+
+
+@app.get("/api/v1/{data}/columns")
```
Same comment, and I would ass that {data} could be misleading with previous route, if it's a `data_type` then change the name for this. 
```suggestion
@app.get("/{data_type}/columns")
```

marcoooo commented on main.py:
```diff
@@ -1,15 +1,77 @@
-@app.get("/")
-def read_root():
-    return {"Hello": "World"}
+@app.get("/api/v1/query/{queryId}/status")
```
```suggestion
@app.get("/query/{queryId}/status")
```

marcoooo commented on main.py:
```diff
@@ -1,15 +1,77 @@
-@app.get("/items/{item_id}")
-def read_item(item_id: int, q: Union[str, None] = None):
-    return {"item_id": item_id, "q": q}
+@app.get("/api/v1/query/{data}/{species}")
```
```suggestion
@app.get("/query/{data_type}/{species}")
```

rohitxsh committed:
feat: handle review comments

rohitxsh committed:
feat: error handling & add dockerfile

rohitxsh committed:
feat: add error codes documentation

rohitxsh committed:
feat

rohitxsh commented:
Changes:
Have split the endpoint for status and preview
Update query endpoint as per content-type content-type: application/hal+json (https://tools.ietf.org/id/draft-kelly-json-hal-01.html)
Preview endpoint now only return 25 rows excluding header
Sort the query string to generate base64 which is used as cache key for redis
Now caching /data and {data_type/columns too
Query endpoint now accepts custom fields too as comma separated values
Redis host and port can be updated by setting env. variable, details added in readme
Added dockerfile, details added in readme
Moved the code to a module
Added logging, added a middleware which logs details like url path, time taken, status code
Added new header X-Correlation-ID which can be used to search for error in logs
Customised the doc generated by FastAPI which can now be used a full fledged documentation for the API endpoints

rohitxsh committed:
feat: log cache hits

rohitxsh committed:
chore: add postman collection

rohitxsh committed:
feat: add description for query params
